tapgarden/planter.go

```diff
@@ -903,7 +1070,86 @@ func (c *ChainPlanter) fundBatch(ctx context.Context, params FundParams) error {
 	return nil
 }
 
-func (c *ChainPlanter) sealBatch(params SealParams) error {
+func (c *ChainPlanter) sealBatch(ctx context.Context, _ SealParams) error {
+	// A batch should exist with 1+ seedlings and be funded at this point.
+	if c.pendingBatch == nil {
+		return fmt.Errorf("no pending batch")
+	}
+
+	if len(c.pendingBatch.Seedlings) == 0 {
+		return fmt.Errorf("no seedlings in batch")
+	}
+
+	if !c.pendingBatch.IsFunded() {
+		return fmt.Errorf("batch is not funded")
+	}
+
+	// filter seedlings to build group key requests; exit early if no
+	// grouped seedlings
+	groupSeedlings, _ := filterSeedlingsWithGroup(c.pendingBatch.Seedlings)
+	if len(groupSeedlings) == 0 {
+		return nil
+	}
+
+	// fetch genesis point and anchor index
+	anchorOutputIndex := uint32(0)
+	if c.pendingBatch.GenesisPacket.ChangeOutputIndex == 0 {
+		anchorOutputIndex = 1
+	}
+
+	genesisPoint := extractGenesisOutpoint(
+		c.pendingBatch.GenesisPacket.Pkt.UnsignedTx,
+	)
+
+	// build genesis TXs
+	groupReqs, genTXs, err := c.buildGroupReqs(
+		genesisPoint, anchorOutputIndex, groupSeedlings,
+	)
+	if err != nil {
+		return fmt.Errorf("unable to build group requests: %w", err)
+	}
+
+	// sign the genesis TXs and validate the authorized asset groups
+	assetGroups := make([]*asset.AssetGroup, 0, len(groupReqs))
+	for i := 0; i < len(groupReqs); i++ {
+		groupKey, err := asset.DeriveGroupKey(
+			c.cfg.GenSigner, genTXs[i], groupReqs[i],
+		)
+		if err != nil {
+			return err
+		}
+
+		protoAsset := groupReqs[i].NewAsset
+		groupedAsset, err := asset.New(
+			protoAsset.Genesis, protoAsset.Amount,
+			protoAsset.LockTime, protoAsset.RelativeLockTime,
+			protoAsset.ScriptKey, groupKey,
+			asset.WithAssetVersion(protoAsset.Version),
+		)
+		if err != nil {
+			return err
+		}
+
+		err = c.cfg.TxValidator.Execute(groupedAsset, nil, nil)
```

Comment:
Where's the witness? Also comments here aren't full sentences above.

Reply:
The witness is a field of `groupKey` above, and we have to re-make the asset with the witness populated to actually validate.